File: mgindb/handler.js
import WebSocket from 'ws'; // Make sure to install 'ws' package via npm

class MginDBClient {
    constructor(host = '127.0.0.1', port = 6446, username = '', password = '') {
        this.uri = `ws://${host}:${port}`;
        this.username = username;
        this.password = password;
        this.websocket = null;
    }

    async connect() {
        this.websocket = new WebSocket(this.uri);

        // Wait for connection to open
        await new Promise((resolve, reject) => {
            this.websocket.on('open', resolve);
            this.websocket.on('error', reject);
        });

        const authData = JSON.stringify({ username: this.username, password: this.password });
        this.websocket.send(authData);

        // Await the first message which should be a welcome message
        const response = await new Promise(resolve => this.websocket.on('message', resolve));
        if (response !== "MginDB server connected... Welcome!") {
            throw new Error("Failed to authenticate: " + response);
        }
    }

    async sendCommand(command) {
        if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
            await this.connect();
        }
        this.websocket.send(command);
        return new Promise(resolve => this.websocket.on('message', resolve));
    }

    async close() {
        if (this.websocket) {
            this.websocket.close();
        }
    }

    set(key, value) {
        return this.sendCommand(`SET ${key} ${value}`);
    }

    indices(action, key = null, value = null) {
        const command = `INDICES ${action}` + (key ? ` ${key}` : '') + (value ? ` ${value}` : '');
        return this.sendCommand(command);
    }

    incr(key, value) {
        return this.sendCommand(`INCR ${key} ${value}`);
    }

    decr(key, value) {
        return this.sendCommand(`DECR ${key} ${value}`);
    }

    delete(key) {
        return this.sendCommand(`DEL ${key}`);
    }

    query(key, queryString = null, options = null) {
        const command = `QUERY ${key}` + (queryString ? ` ${queryString}` : '') + (options ? ` ${options}` : '');
        return this.sendCommand(command);
    }

    count(key) {
        return this.sendCommand(`COUNT ${key}`);
    }

    schedule(action, cronOrKey = null, command = null) {
        const cmd = `SCHEDULE ${action}` + (cronOrKey ? ` ${cronOrKey}` : '') + (command ? ` ${command}` : '');
        return this.sendCommand(cmd);
    }

    sub(key) {
        return this.sendCommand(`SUB ${key}`);
    }

    unsub(key) {
        return this.sendCommand(`UNSUB ${key}`);
    }
}

export default MginDBClient;

File: mgindb/handler.cs
using System;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using WebSocketSharp;

public class MginDBClient
{
    private readonly string _username;
    private readonly string _password;
    private readonly WebSocket _webSocket;

    public MginDBClient(string host = "127.0.0.1", int port = 6446, string username = "", string password = "")
    {
        _username = username;
        _password = password;
        _webSocket = new WebSocket($"ws://{host}:{port}");
        _webSocket.OnMessage += (sender, e) => OnMessageReceived(e.Data);
    }

    private TaskCompletionSource<string> _responseTcs;

    private void OnMessageReceived(string message)
    {
        _responseTcs?.SetResult(message);
    }

    public async Task ConnectAsync()
    {
        var tcs = new TaskCompletionSource<bool>();

        _webSocket.OnOpen += (sender, e) =>
        {
            var authData = new { username = _username, password = _password };
            _webSocket.Send(JsonConvert.SerializeObject(authData));
            tcs.SetResult(true);
        };

        _webSocket.OnMessage += (sender, e) =>
        {
            if (e.Data == "MginDB server connected... Welcome!")
            {
                tcs.SetResult(true);
            }
            else
            {
                tcs.SetException(new Exception("Failed to authenticate: " + e.Data));
            }
        };

        _webSocket.ConnectAsync();
        await tcs.Task;
    }

    private async Task<string> SendCommandAsync(string command)
    {
        if (!_webSocket.IsAlive)
        {
            await ConnectAsync();
        }

        _responseTcs = new TaskCompletionSource<string>();
        _webSocket.Send(command);
        return await _responseTcs.Task;
    }

    public async Task<string> Set(string key, string value)
    {
        return await SendCommandAsync($"SET {key} {value}");
    }

    public async Task<string> Indices(string action, string key = null, string value = null)
    {
        return await SendCommandAsync($"INDICES {action} {(key ?? "")} {(value ?? "")}".Trim());
    }

    public async Task<string> Incr(string key, string value)
    {
        return await SendCommandAsync($"INCR {key} {value}");
    }

    public async Task<string> Decr(string key, string value)
    {
        return await SendCommandAsync($"DECR {key} {value}");
    }

    public async Task<string> Delete(string key)
    {
        return await SendCommandAsync($"DEL {key}");
    }

    public async Task<string> Query(string key, string queryString = null, string options = null)
    {
        return await SendCommandAsync($"QUERY {key} {(queryString ?? "")} {(options ?? "")}".Trim());
    }

    public async Task<string> Count(string key)
    {
        return await SendCommandAsync($"COUNT {key}");
    }

    public async Task<string> Schedule(string action, string cronOrKey = null, string command = null)
    {
        return await SendCommandAsync($"SCHEDULE {action} {(cronOrKey ?? "")} {(command ?? "")}".Trim());
    }

    public async Task<string> Sub(string key)
    {
        return await SendCommandAsync($"SUB {key}");
    }

    public async Task<string> Unsub(string key)
    {
        return await SendCommandAsync($"UNSUB {key}");
    }

    public void Close()
    {
        _webSocket.Close();
    }
}
